src/util/strencodings.h

```diff
@@ -369,4 +370,44 @@ std::string Capitalize(std::string str);
  */
 std::optional<uint64_t> ParseByteUnits(std::string_view str, ByteUnit default_multiplier);
 
+namespace util {
+/** consteval version of HexDigit() without the lookup table. */
+consteval uint8_t ConstevalHexDigit(const char c)
+{
+    if (c >= '0' && c <= '9') return c - '0';
+    if (c >= 'a' && c <= 'f') return c - 'a' + 0xa;
+
+    throw "Only lowercase hex digits are allowed, for consistency";
+}
+
+/** Converts from hex string literal to byte array at compile time.
+ *
+ * @note It may be preferable to use Vec(HexLiteral(...)) instead of
+ * HexLiteral() to save stack space when declaring a local variable, if the hex
+ * string is large. Alternately the variable could be declared constexpr to
+ * avoid using stack space.
+ */
+template <typename Byte = std::byte, size_t N>
+consteval std::array<Byte, N / 2> HexLiteral(const char (&hex_str)[N])
```

Comment:
Given that we already seem to be using custom string overloads (e.g. https://github.com/bitcoin/bitcoin/blob/master/src/script/miniscript.cpp#L87), would it maybe make the usages less noisy if we tried something like:
```C++
consteval auto operator""_hex(const char* hex_str, const std::size_t len)
```
to be able to write `"0000deadbeef0000"_hex` instead of `HexLiteral<uint8_t>("0000deadbeef0000")`?

I haven't spent a lot of time with this to make it work since I'm not yet sure it's a good idea.

Reply:
I'm not advocating for it, but I think we can get it to compile by something like:
```C++
consteval auto operator"" _hex(const char* hex_str, size_t _)
{
    constexpr std::size_t len = sizeof(hex_str);
    static_assert(len % 2 == 0, "Hex string must have an even number of characters");

    if (hex_str[len - 1] != '\0') throw "null terminator required";
    std::array<std::byte, len / 2> rv{};
    size_t i = 0;
    for (auto& elem: rv) {
        auto hi = ConstevalHexDigit(hex_str[i++]) << 4;
        elem = static_cast<std::byte>(hi | ConstevalHexDigit(hex_str[i++]));
    }
    return rv;
}
```

but it's kinda' ugly and might not resolve our problems. Please resolve the comment if this isn't helpful.

Reply:
I tried before your suggestion and then tried again when you suggested on all 3 compilers and none worked. Which I stated.

Then you pasted a longer example. Have you tested and got anything remotely close to that working on any compiler, C++ or other language? :)

Resolving.

Reply:
The above seemed to be working (that's why I suggested it originally) - but I'm not in love with this version either - thanks for considering.

Reply:
I see that my example above wasn't working with GCC and had a stupid `sizeof` bug.

Here's an alternative using [string literal operator template](https://en.cppreference.com/w/cpp/language/user_literal#Literal_operators) which seems to be working correctly for me:
```C++
template<size_t N>
struct _Hex {
    std::array<std::byte, N / 2> hex;
    consteval _Hex(const char (&hex_str)[N]) requires (N % 2 == 1) {
        if (hex_str[N - 1]) throw "null terminator required";
        for (std::size_t i = 0; i < hex.size(); ++i) {
            hex[i] = static_cast<std::byte>(
                (ConstevalHexDigit(hex_str[2 * i]) << 4) |
                 ConstevalHexDigit(hex_str[2 * i + 1]));
        }
    }
};

template<_Hex str>
consteval auto operator"" _hex() { return str.hex; }

template<_Hex str>
constexpr auto operator"" _hex_v() { return Vec(str.hex); }
```
used as
```C++
auto hex_array = "9caaf126043eb5"_hex;
auto hex_vector = "9caaf126043eb5"_hex_v;
```

<details>
<summary>string_hex test</summary>

```C++
diff --git a/src/test/miner_tests.cpp b/src/test/miner_tests.cpp
--- a/src/test/miner_tests.cpp	(revision 424f8f4ef5042adf64764c65e85d5b25faf60e66)
+++ b/src/test/miner_tests.cpp	(date 1724169991324)
@@ -29,6 +29,7 @@
 
 using node::BlockAssembler;
 using node::CBlockTemplate;
+using util::ConstevalHexDigit;
 using util::HexLiteral;
 using util::Vec;
 
@@ -660,4 +661,55 @@
     TestPrioritisedMining(scriptPubKey, txFirst);
 }
 
+/////////
+
+template<size_t N>
+struct _Hex {
+    std::array<std::byte, N / 2> hex;
+    consteval _Hex(const char (&hex_str)[N]) requires (N % 2 == 1) {
+        if (hex_str[N - 1]) throw "null terminator required";
+        for (std::size_t i = 0; i < hex.size(); ++i) {
+            hex[i] = static_cast<std::byte>(
+                (ConstevalHexDigit(hex_str[2 * i]) << 4) |
+                ConstevalHexDigit(hex_str[2 * i + 1]));
+        }
+    }
+};
+
+template<_Hex str>
+consteval auto operator"" _hex() { return str.hex; }
+
+template<_Hex str>
+constexpr auto operator"" _hex_v() { return Vec(str.hex); }
+
+/////////
+
+template <typename Byte = std::byte, size_t N>
+consteval std::array<Byte, N / 2> HexLiteral(const char (&hex_str)[N]) requires (N % 2 == 1) {
+    if (hex_str[N - 1] != '\0') throw "null terminator required";
+    std::array<Byte, N / 2> rv{};
+    size_t i = 0;
+    for (auto& elem : rv) {
+        auto hi = ConstevalHexDigit(hex_str[i++]) << 4;
+        elem = static_cast<Byte>(hi | ConstevalHexDigit(hex_str[i++]));
+    }
+    return rv;
+}
+
+BOOST_AUTO_TEST_CASE(string_hex)
+{
+    auto original = HexLiteral("04678afdb0fe5548271967f1a67130b7105cd6a828e03909a67962e0ea1f61deb649f6bc3f4cef38c4f35504e51ec112de5c384df7ba0b8d578a4c702b6bf11d5f");
+    static_assert(std::is_same_v<decltype(original), std::array<std::byte, 65>>);
+    std::cout << HexStr(original) << std::endl;
+
+    auto hex_value = "04678afdb0fe5548271967f1a67130b7105cd6a828e03909a67962e0ea1f61deb649f6bc3f4cef38c4f35504e51ec112de5c384df7ba0b8d578a4c702b6bf11d5f"_hex;
+    static_assert(std::is_same_v<decltype(hex_value), std::array<std::byte, 65>>);
+    std::cout << HexStr(hex_value) << std::endl;
+
+    auto hex_value2 = "9caaf126043eb5"_hex_v;
+    static_assert(std::is_same_v<decltype(hex_value2), std::vector<std::byte>>);
+    assert(hex_value2.size() == 7);
+    std::cout << HexStr(hex_value2) << std::endl;
+}
+
 BOOST_AUTO_TEST_SUITE_END()
```

</details>

See: https://godbolt.org/z/qPz4Ex8dT

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30377#discussion_r1723595926

> Here's an alternative using [string literal operator template](https://en.cppreference.com/w/cpp/language/user_literal#Literal_operators) which seems to be working correctly for me:

Wow, that looks great! Assuming this works on compilers we need, it looks like a very nice solution that completely avoids problems I was [struggling with](https://github.com/bitcoin/bitcoin/pull/30377#issuecomment-2291521586) trying to get a function parameter to act like a template parameter.

Reply:
That is awesome! Didn't realize I had to *enable* C++20 support on Godbolt. :man_facepalming:  Thank you for persisting @paplorinc! Will experiment with this.

Reply:
Even though the prior version had many ACKs, this reaches another level of awesome.
Initial version using `""_hex` now pushed in daba1a25a62e72e9797a134c6377d17a9274a25f with @l0rinc as main author of that commit! :tada: 

Resolving this specific thread.